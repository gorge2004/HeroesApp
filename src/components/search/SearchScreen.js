import { useNavigate, useLocation } from "react-router-dom";
import queryString from "query-string";
import { useForm } from "../../hooks/useForm";
import { getHeoresByName } from "../../selectorts/getHeroByName";
import HeroCard from "../hero/HeroCard";
import { useMemo } from 'react';

function SearchScreen() {
  const navigate = useNavigate();
  const location = useLocation();
  const { q = "" } = queryString.parse(location.search);
  const [values, handleInputChange, reset] = useForm({
    searchTerm: q,
  });

  const { searchTerm } = values;
  let heroesFiltered = useMemo(()=> getHeoresByName(q), [q]) ;
  const handleSubmit = (evt) => {
    evt.preventDefault();
    if (searchTerm.trim().length > 1) {
      navigate(`?q=${searchTerm}`);
      heroesFiltered = getHeoresByName(searchTerm);
    }
  };
  return (
    <>
      <h1>Busquedas</h1>
      <div className="row">
        <div className="col-5">
          <h4>Buscar</h4>
          <hr />
          <form onSubmit={handleSubmit}>
            <input
              className="form-control"
              type="text"
              placeholder="Buscar un heroe"
              name="searchTerm"
              autoComplete="off"
              value={searchTerm}
              onChange={handleInputChange}
            />
            <button className="btn btn-outline-primary mt-1">Buscar</button>
          </form>
        </div>

        <div className="col-7">
          <h4>Resultados</h4>
          <hr />

          { q.length == 0 ? (
            <div className="alert alert-info">
              Buscar un heroe.
            </div>
          )
            :
            ( heroesFiltered.length === 0 && ( <div className="alert alert-danger">No hay resultados: {q}</div>))
        }
          {heroesFiltered.map((heroe) => (
            <HeroCard key={heroe.id} {...heroe} />
          ))}
        </div>
      </div>
    </>
  );
}

export default SearchScreen;
