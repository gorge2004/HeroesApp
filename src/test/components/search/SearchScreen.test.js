import SearchScreen from "../../../components/search/SearchScreen";
import { mount } from "enzyme";
import { AuthContext } from "../../../auth/authContext";
import { MemoryRouter } from "react-router-dom";
//estoy reescribinedo todo los metodos, de la libreria y solo necesito modificar el useNavigation
//uso jest.requireActual traerme todos sus valores y no rompa otros test y modificar lo que necesito

const mockNavigate = jest.fn();
jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => mockNavigate,
}));

describe("testing_SearchScreen_component", () => {
  const contextValue = {
    user: {
      name: "John",
      logged: true,
    },
  };
  test("should_render_SearchScreen", () => {
    const wrapper = mount(
      <AuthContext.Provider value={contextValue}>
        <MemoryRouter initialEntries={["/search"]}>
          <SearchScreen />
        </MemoryRouter>
      </AuthContext.Provider>
    );
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(".alert-info").text().trim()).toBe("Buscar un heroe.");
  });

  test("should_render_component_with_batman", () => {
    const q = "batman";
    const wrapper = mount(
      <AuthContext.Provider value={contextValue}>
        <MemoryRouter initialEntries={[`/search?q=${q}`]}>
          <SearchScreen />
        </MemoryRouter>
      </AuthContext.Provider>
    );
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find("input").prop("value").trim()).toBe(q);
  });
  test("should_render_a_error_Element_if_heros_name_not exists", () => {
    const q = "batmandasdasdasdasdasdas";
    const wrapper = mount(
      <AuthContext.Provider value={contextValue}>
        <MemoryRouter initialEntries={[`/search?q=${q}`]}>
          <SearchScreen />
        </MemoryRouter>
      </AuthContext.Provider>
    );
    expect(wrapper.find(".alert-danger").exists()).toBeTruthy();
    expect(wrapper.find(".alert-danger").text().trim()).toBe(
      `No hay resultados: ${q}`
    );
  });

  test("should_call_navigate", () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={[`/search`]}>
        <SearchScreen />
      </MemoryRouter>
    );
    wrapper
      .find("input")
      .simulate("change", { target: { name: "searchTerm", value: "batman" } });
    wrapper.find("form").prop("onSubmit")({ preventDefault: jest.fn() });
    expect(mockNavigate).toHaveBeenCalledWith("?q=batman");
  });
});
