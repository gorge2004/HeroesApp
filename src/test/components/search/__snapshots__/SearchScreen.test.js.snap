// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing_SearchScreen_component should_render_SearchScreen 1`] = `
Array [
  <h1>
    Busquedas
  </h1>,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Buscar
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchTerm"
          onChange={[Function]}
          placeholder="Buscar un heroe"
          type="text"
          value=""
        />
        <button
          className="btn btn-outline-primary mt-1"
        >
          Buscar
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Resultados
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        Buscar un heroe.
      </div>
    </div>
  </div>,
]
`;

exports[`testing_SearchScreen_component should_render_component_with_batman 1`] = `
Array [
  <h1>
    Busquedas
  </h1>,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Buscar
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchTerm"
          onChange={[Function]}
          placeholder="Buscar un heroe"
          type="text"
          value="batman"
        />
        <button
          className="btn btn-outline-primary mt-1"
        >
          Buscar
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Resultados
      </h4>
      <hr />
      <div
        className="col animate__animated animate__fadeIn"
      >
        <div
          className="card"
        >
          <div
            className="row no-gutters"
          >
            <div
              className="col-4"
            >
              <img
                alt="Batman"
                className="card-img"
                src="/assets/dc-batman.jpg"
              />
            </div>
            <div
              className="col-8"
            >
              <div
                className="card-body"
              >
                <h5
                  className="card-title"
                >
                  Batman
                </h5>
                <p
                  className="card-name"
                >
                  Bruce Wayne
                </p>
                <p
                  className="card-text"
                >
                  <a
                    href="/hero/dc-batman"
                    onClick={[Function]}
                  >
                    Mas...
                  </a>
                </p>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
]
`;
